### PR TITLE
[ci] dowgrade gcc version for Linux on Azure

### PR DESCRIPTION
**Pull request by StrikerRUS**
Refer to #1708.

### CODE REVIEW

**Comment by StrikerRUS** [comment]
GPU test fails. I suppose that the reason is in these lines:
```
2018-09-30T17:18:46.9243394Z Preparing to unpack .../ocl-icd-libopencl1_2.2.8-1_amd64.deb ...
2018-09-30T17:18:46.9646898Z Unpacking ocl-icd-libopencl1:amd64 (2.2.8-1) ...
2018-09-30T17:18:47.1272916Z Selecting previously unselected package opencl-headers.
2018-09-30T17:18:47.1380157Z Preparing to unpack .../opencl-headers_2.0~svn32091-2_all.deb ...
2018-09-30T17:18:47.1548332Z Unpacking opencl-headers (2.0~svn32091-2) ...
2018-09-30T17:18:47.2831063Z Selecting previously unselected package ocl-icd-opencl-dev:amd64.
2018-09-30T17:18:47.2923796Z Preparing to unpack .../ocl-icd-opencl-dev_2.2.8-1_amd64.deb ...
2018-09-30T17:18:47.3099881Z Unpacking ocl-icd-opencl-dev:amd64 (2.2.8-1) ...
2018-09-30T17:18:47.3866224Z Processing triggers for libc-bin (2.23-0ubuntu10) ...
2018-09-30T17:18:47.5810345Z Setting up ocl-icd-libopencl1:amd64 (2.2.8-1) ...
2018-09-30T17:18:47.6531984Z Setting up opencl-headers (2.0~svn32091-2) ...
2018-09-30T17:18:47.7040672Z Setting up ocl-icd-opencl-dev:amd64 (2.2.8-1) ...
---> 2018-09-30T17:18:47.7545016Z Processing triggers for libc-bin (2.23-0ubuntu10) ... <---
```

While installing OpenCL libraries `libc` is being updated and becomes incompatible with g++-4.8. It's my assumption.

@guolinke Since, we don't provide any artifacts from GPU test publicly, what do you think about the following?

```
elif [[ $AZURE == "true" ]] && [[ $COMPILER == "gcc" ]] && [[ $TASK != "GPU" ]]; then
  <downgrade gcc>
```

**Comment by guolinke** [comment]
@StrikerRUS yeah, that is okay

File: .ci/setup.sh [hunk 1]
```diff
@@ -22,21 +22,27 @@ if [[ $OS_NAME == "macos" ]]; then
         wget -O conda.sh https://repo.continuum.io/miniconda/Miniconda${PYTHON_VERSION:0:1}-latest-MacOSX-x86_64.sh
     fi
 else  # Linux
-    if [[ $AZURE == "true" ]] && [[ $COMPILER == "clang" ]]; then
-        update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-6.0 100
-        update-alternatives --install /usr/bin/clang clang /usr/bin/clang-6.0 100
+    if [[ $AZURE == "true" ]]; then
         sudo apt-get update
```
Comment (guolinke) [inline .ci/setup.sh]: may be we should update for travis as well, since line 48 will install ocl-icd-opencl-dev

**Comment by guolinke** [comment]
@StrikerRUS 
It seems the build binaries still require glibc 2.23:
```
guoke@test-vm3:~$ ldd lib_lightgbm.so
./lib_lightgbm.so: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.23' not found (required by ./lib_lightgbm.so)
        linux-vdso.so.1 =>  (0x00007ffc472ae000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f8ecbcc6000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f8ecb9c0000)
        libgomp.so.1 => /usr/lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f8ecb7b1000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f8ecb59b000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f8ecb37d000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f8ecafb4000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f8ecc445000)
```
So i add some fixes .

File: .ci/setup.sh (hunk 1 above)
Comment (StrikerRUS) [reply .ci/setup.sh]: I tested without `sudo apt-get update` at Travis - worked well. But if it's not significantly enlarge the test time, lets leave this.

File: CMakeLists.txt [hunk 2]
```diff
@@ -17,6 +17,10 @@ if(APPLE)
     OPTION(APPLE_OUTPUT_DYLIB "Output dylib shared library" OFF)
 endif()
 
+if((CMAKE_CXX_COMPILER_ID STREQUAL "GNU") AND (UNIX) AND (NOT APPLE))
+  ADD_DEFINITIONS(-DGCC)
+endif()
```
Comment (StrikerRUS) [inline CMakeLists.txt]: Maybe it's better to move this somewhere lower to other definitions?

Comment (guolinke) [reply CMakeLists.txt]: sorry, I didn't understand it, can you describe more details ?

Comment (StrikerRUS) [reply CMakeLists.txt]: I mean, move these lines lower in the file. For example, somewhere around here
https://github.com/Microsoft/LightGBM/blob/606eeefeb8e1f217a0f91fa0f11137125942b549/CMakeLists.txt#L66-L68
or here
https://github.com/Microsoft/LightGBM/blob/606eeefeb8e1f217a0f91fa0f11137125942b549/CMakeLists.txt#L118-L119
where other `ADD_DEFINITIONS` are set.

File: include/LightGBM/meta.h [hunk 3]
```diff
@@ -62,6 +62,10 @@ typedef void(*AllgatherFunction)(char* input, comm_size_t input_size, const comm
 #define __func__ __FUNCTION__
 #endif
 
+#ifdef GCC
+__asm__(".symver lgamma, lgamma@GLIBC_2.2.5");
```
[the remark this answers is not in the document]
Comment (superbobry) [reply include/LightGBM/meta.h]: Hmm, why is this symver needed?

[the remark this answers is not in the document]
Comment (guolinke) [reply include/LightGBM/meta.h]: `lgamma` will introduce the glibc2.23 dependency.

**Comment by guolinke** [comment]
@StrikerRUS what do you think about the current fix for `lgamma` ?

**Comment by StrikerRUS** [comment]
@guolinke To be honest, I'm not familiar with SymbolVersioning, but I think it's rather hacky solution and intuitively I can't say that I like it. I'm afraid that it will cause some troubles.

I'd better prefer this your proposal:
>maybe a better solution is implementing a lgamma by ourself.

Can we just copy-paste this function from the sources?

File: include/LightGBM/meta.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (superbobry) [reply include/LightGBM/meta.h]: This sounds a bit odd. I've checked the glibc [changelog](https://sourceware.org/git/?p=glibc.git;a=tree;f=ChangeLog.old;h=9359206ab5f94d23170eb0560bae2c9d642eb43b;hb=HEAD) and it seems `lgamma` have been there for quite a while.

[the remark this answers is not in the document]
Comment (guolinke) [reply include/LightGBM/meta.h]: I think lgamma have a new implementation in glibc2.23. and result in this new dependency.

**Comment by guolinke** [comment]
@StrikerRUS I just notice we don't need the lgamma. The new fix is pushed.

**Comment by superbobry** [comment]
What is the glibc version used by GCC 4.8? Could you mention it in the PR description?

**Comment by StrikerRUS** [comment]
@guolinke 
>I just notice we don't need the lgamma

Wow, it's great! This fix is for sure much better than symbol versioning. 😃

**Comment by StrikerRUS** [comment]
@guolinke I've added `... || exit -1` to prevent silent usage of inappropriate gcc.

**Comment by guolinke** [comment]
@superbobry 
```
guoke@rdnn-gpu-nv-01:~$ objdump -T lib_lightgbm.so | fgrep GLIBC_
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 __printf_chk
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __errno_location
0000000000000000      DO *UND*  0000000000000000  GLIBC_2.2.5 stdout
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 bind
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 ceil
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 socket
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 __vfprintf_chk
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 nanosleep
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 accept
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 log
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 recv
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 __cxa_finalize
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strlen
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 memcmp
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 send
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 memset
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_create
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 __sprintf_chk
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 __snprintf_chk
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 vsnprintf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __assert_fail
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 nextafter
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_mutex_unlock
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strtod
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sqrt
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.14  memcpy
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 connect
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 __strcpy_chk
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __cxa_atexit
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pow
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fclose
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 inet_ntop
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.4   __stack_chk_fail
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 listen
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fflush
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fopen
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 exp
0000000000000000      DO *UND*  0000000000000000  GLIBC_2.2.5 stderr
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 __vsprintf_chk
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strtol
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_mutex_lock
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fread
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 log2f
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 log2
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 memmove
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 __fprintf_chk
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 __pthread_key_create
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3   __tls_get_addr
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3   freeifaddrs
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 setsockopt
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sqrtf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3   getifaddrs
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 inet_pton
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fwrite
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 close
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 logf
```

**Comment by guolinke** [comment]
yeah, refer to#1727

**Pull request merged**